### PR TITLE
=htc #1610 fix "Cannot push port ... twice" in NewHostConnectionPool

### PR DESCRIPTION
**Pull request by jrudolph**

### CODE REVIEW

File: akka-http-core/src/main/scala/akka/http/impl/engine/client/pool/NewHostConnectionPool.scala [hunk 1]
```diff
@@ -301,7 +300,7 @@ private[client] object NewHostConnectionPool {
           }
           def isClosed: Boolean = requestOut.isClosed || responseIn.isClosed
 
-          def onPush(): Unit = {
+          def onPush(): Unit = try {
```
Comment (raboof) [inline akka-http-core/src/main/scala/akka/http/impl/engine/client/pool/NewHostConnectionPool.scala]: What does `try` achieve here?

**Comment by raboof** [comment]
Refs #1610

File: akka-http-core/src/main/scala/akka/http/impl/engine/client/pool/NewHostConnectionPool.scala [hunk 2]
```diff
@@ -344,11 +343,23 @@ private[client] object NewHostConnectionPool {
...
             withSlot(_.debug("Connection cancelled"))
 
+          /** Helper that makes sure requestOut is pulled before pushing */
+          private def emitRequest(request: HttpRequest): Unit =
+            if (requestOut.isAvailable) requestOut.push(request)
+            else
+              requestOut.setHandler(new OutHandler {
+                def onPull(): Unit = {
+                  requestOut.push(request)
+                  requestOut.setHandler(connection) // implicit assumption is that connection was the handler before
```
Comment (raboof) [inline akka-http-core/src/main/scala/akka/http/impl/engine/client/pool/NewHostConnectionPool.scala]: Perhaps we should make this assumption explicit and fail fast by moving this inner `OutHandler` class to its own type and checking this existing handler is not of that type before doing `requestOut.setHandler(new ...)` above?

File: akka-http-core/src/main/scala/akka/http/impl/engine/client/pool/NewHostConnectionPool.scala (hunk 1 above)
Comment (jrudolph) [reply akka-http-core/src/main/scala/akka/http/impl/engine/client/pool/NewHostConnectionPool.scala]: Oops, looks like a left-over. Weird that it compiles.

File: akka-http-core/src/main/scala/akka/http/impl/engine/client/pool/NewHostConnectionPool.scala (hunk 2 above)
Comment (jrudolph) [reply akka-http-core/src/main/scala/akka/http/impl/engine/client/pool/NewHostConnectionPool.scala]: There's no way to get the current handler for `SubSourceOutlet` (see https://github.com/akka/akka/issues/24158 for other inconsistencies between `SubSourceOutlet` and regular ones). I'll update the comment.

**Comment by jrudolph** [comment]
> Would be nice to have a test that more reliably reproduces this than #1610, but seems reasonable!

It is completely reproducible on Scala 2.12.

**Comment by raboof** [comment]
Needs MiMa exclusions for these internals:
```
[info] akka-http-core: found 2 potential binary incompatibilities while checking against com.typesafe.akka:akka-http-core_2.11:10.0.11 
[error]  * method onShutdown(akka.http.impl.engine.client.pool.SlotContext)Unit in trait akka.http.impl.engine.client.pool.SlotState#BusyState is present only in current version
[error]    filter with: ProblemFilters.exclude[ReversedMissingMethodProblem]("akka.http.impl.engine.client.pool.SlotState#BusyState.onShutdown")
[error]  * synthetic method akka$http$impl$engine$client$pool$SlotState$BusyState$$super$onShutdown(akka.http.impl.engine.client.pool.SlotContext)Unit in trait akka.http.impl.engine.client.pool.SlotState#BusyState is present only in current version
[error]    filter with: ProblemFilters.exclude[ReversedMissingMethodProblem]("akka.http.impl.engine.client.pool.SlotState#BusyState.akka$http$impl$engine$client$pool$SlotState$BusyState$$super$onShutdown")
```

**Comment by jrudolph** [comment]
Triggered nightlies to see if it is fixed for good.

**Pull request merged**